internal/xds/xdsdependencymanager/xds_dependency_manager.go

```diff
@@ -0,0 +1,241 @@
+	matchVh := xdsresource.FindBestMatchingVirtualHost(m.dataplaneAuthority, update.VirtualHosts)
+	if matchVh == nil {
+		m.watcher.OnError(status.Errorf(codes.Unavailable, "Could not find VirtualHost for %q", m.dataplaneAuthority))
+		return
+	}
+	m.currentRouteConfig = update
+	m.currentVirtualHost = matchVh
+	m.maybeSendUpdate()
+}
+
+// Only executed in the context of a serializer callback.
+func (m *DependencyManager) onListenerResourceUpdate(update *xdsresource.ListenerUpdate) {
+	if m.logger.V(2) {
+		m.logger.Infof("Received update for Listener resource %q: %v", m.ldsResourceName, pretty.ToJSON(update))
+	}
+
+	m.currentListenerUpdate = update
+
+	if update.InlineRouteConfig != nil {
+		// If there was a previous route config watcher because of a non-inline
+		// route configuration, cancel it.
+		m.rdsResourceName = ""
+		if m.routeConfigWatcher != nil {
+			m.routeConfigWatcher.stop()
+			m.routeConfigWatcher = nil
+		}
+
+		m.applyRouteConfigUpdate(*update.InlineRouteConfig)
+		return
+	}
+
+	// We get here only if there was no inline route configuration.
+
+	// If the route config name has not changed, send an update with existing
+	// route configuration and the newly received listener configuration.
+	if m.rdsResourceName == update.RouteConfigName {
+		m.maybeSendUpdate()
+		return
+	}
+
+	// If the route config name has changed, cancel the old watcher and start a
+	// new one. At this point, since we have not yet resolved the new route
+	// config name, we don't send an update to the channel, and therefore
+	// continue using the old route configuration (if received) until the new
+	// one is received.
+	m.rdsResourceName = update.RouteConfigName
+	if m.routeConfigWatcher != nil {
+		m.routeConfigWatcher.stop()
+		m.currentVirtualHost = nil
+	}
+	m.routeConfigWatcher = newRouteConfigWatcher(m.rdsResourceName, m)
+}
+
+// Only executed in the context of a serializer callback.
+func (m *DependencyManager) onListenerResourceError(err error) {
+	if m.logger.V(2) {
+		m.logger.Infof("Received resource error for Listener resource %q: %v", m.ldsResourceName, err)
+	}
+
+	if m.routeConfigWatcher != nil {
+		m.routeConfigWatcher.stop()
+	}
+	m.rdsResourceName = ""
+	m.currentVirtualHost = nil
+	m.routeConfigWatcher = nil
+	m.watcher.OnError(status.Errorf(codes.Unavailable, "Listener resource error : %v", err))
+}
+
+// Only executed in the context of a serializer callback.
+func (m *DependencyManager) onListenerResourceAmbientError(err error) {
+	m.logger.Warningf("Listener resource ambient error: %v", err)
+}
+
+// Only executed in the context of a serializer callback.
+func (m *DependencyManager) onRouteConfigResourceUpdate(resourceName string, update xdsresource.RouteConfigUpdate) {
+	if m.logger.V(2) {
+		m.logger.Infof("Received update for RouteConfiguration resource %q: %v", resourceName, pretty.ToJSON(update))
+	}
+
+	if m.rdsResourceName != resourceName {
+		// Drop updates from canceled watchers.
+		return
+	}
```

Comment:
I'm guessing we are going need code like this for cluster and endpoint watchers as well. Can we make this part of the watcher instead?